src/solver/reduce/forward_sum.cpp

```diff
@@ -0,0 +1,334 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/reduce/solvers.hpp>
+
+#include <miopen/reduce/invoke_params.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/sum.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/target_properties.hpp>
+
+#define LOCAL_SIZE 256
+
+namespace miopen {
+
+namespace solver {
+
+namespace reduce {
+
+bool IsNotLastDim(const miopen::reduce::ProblemDescription& problem)
+{
+    if((problem.GetDim() == problem.GetXDesc().GetLengths().size() - 1))
+    {
+        MIOPEN_THROW(miopenStatusBadParm, "SumForward: Reduce last dim not supported.");
+    }
+    return true;
+}
+
+bool IsImprovementOverROCm(const ExecutionContext& context,
+                           const miopen::reduce::ProblemDescription& problem)
+{
+    auto xdims = problem.GetXDesc().GetLengths();
+    auto ydims = problem.GetYDesc().GetLengths();
+    auto dim   = problem.GetDim();
+
+    auto reduce_size = xdims[dim];
+    auto output_numel =
+        std::accumulate(ydims.begin(), ydims.end(), 1ULL, std::multiplies<size_t>());
+
+    // At least 4 WGs per one CU
+    auto reqd_work_item_cnt =
+        static_cast<size_t>(256 * context.GetStream().GetMaxComputeUnits() * 4);
+
+    bool is_num_work_item_enough = (output_numel > reqd_work_item_cnt);
+    bool is_parallelism_enough   = (output_numel * reduce_size > reqd_work_item_cnt);
+
+    if(!is_num_work_item_enough && is_parallelism_enough)
+    {
+        // It's large enough to parallelization, but calling the kernel twice is overhead.
+        // For cases smaller than this, ROCm pytorch performed better.
+        bool is_improvement_ROCm = (output_numel * reduce_size < reqd_work_item_cnt * 64);
+        // But the reduce size is small, MIOpen HIP performs better.
+        bool is_reduce_large = (reduce_size > 64);
+
+        if(is_improvement_ROCm && is_reduce_large)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "SumForward: Performance cannot be guaranteed.");
```

Comment:
Why do we fail here? Either we should return `false` or `true` and put something into the logs.

Reply:
fixed it